ambry-frontend/src/main/java/com/github/ambry/frontend/TtlUpdateHandler.java

```diff
@@ -143,17 +158,34 @@ private Callback<String> idConverterCallback() {
     private Callback<Void> securityPostProcessRequestCallback(BlobId blobId) {
       return buildCallback(metrics.updateBlobTtlSecurityPostProcessRequestMetrics, result -> {
         String serviceId = RestUtils.getHeader(restRequest.getArgs(), RestUtils.Headers.SERVICE_ID, true);
-        router.updateBlobTtl(blobId.getID(), serviceId, Utils.Infinite_Time, routerCallback(),
+        router.updateBlobTtl(blobId.getID(), serviceId, Utils.Infinite_Time, routerCallback(blobId),
             QuotaUtils.buildQuotaChargeCallback(restRequest, quotaManager, false));
       }, restRequest.getUri(), LOGGER, finalCallback);
     }
 
     /**
      * After {@link Router#updateBlobTtl} finishes, call {@link SecurityService#processResponse}.
+     * @param blobId the {@link BlobId} to update
      * @return a {@link Callback} to be used with {@link Router#updateBlobTtl}.
      */
-    private Callback<Void> routerCallback() {
+    private Callback<Void> routerCallback(BlobId blobId) {
       return buildCallback(metrics.updateBlobTtlRouterMetrics, result -> {
+        if (RequestPath.matchesOperation(blobIdStr, Operations.NAMED_BLOB)) {
+          // Set the named blob state to be 'READY' after the Ttl update succeed
+          if (!restRequest.getArgs().containsKey(NAMED_BLOB_VERSION)) {
+            throw new RestServiceException(
+                "Internal key " + NAMED_BLOB_VERSION + " is required in Named Blob TTL update callback!",
+                RestServiceErrorCode.InternalServerError);
+          }
+          long namedBlobVersion = (long) restRequest.getArgs().get(NAMED_BLOB_VERSION);
+          NamedBlobPath namedBlobPath = NamedBlobPath.parse(blobIdStr, restRequest.getArgs());
+          NamedBlobRecord record = new NamedBlobRecord(namedBlobPath.getAccountName(), namedBlobPath.getContainerName(),
+              namedBlobPath.getBlobName(), blobId.getID(), Utils.Infinite_Time, namedBlobVersion);
+          namedBlobDb.updateBlobTtlAndStateToReady(record).get();
```

Comment:
nit: do you think we should put the rest of the logic as a callback to the CompletableFuture returned by `updateBlobTtlAndStateToReady`?

Reply:
I've updated it, but just wondering what's your concern about that. 
Is it because you think it's more clear if we chain the flow with callback?

Reply:
database call can take some time, and we are blocking it here, by changing it to callback chain (completable future chain), we are not blocking it.